File: drivers/javascript/meshdb.mjs
// meshdb HTTP driver — Node 18+ (uses built-in fetch), zero dependencies, streaming reads.
//
//   import { Client } from "./meshdb.mjs";
//   const db = new Client("http://localhost:4680", { user: "app", secret: "s3cret" });
//   for await (const row of db.query("SELECT id, v FROM t WHERE id > ?", { params: [5] }))
//     console.log(row.id, row.v);
//   await db.execute("INSERT INTO t VALUES (?, ?)", { params: [1, "a"] });
//
// query() is an async generator: rows arrive one at a time, so a million-row result costs the
// driver almost nothing, matching the gateway's streaming. Auth is sent as
// `Authorization: Bearer base64(user:secret)` — the programmatic scheme (no browser prompt).
// Over a plaintext gateway the credential is exposed; use TLS on any untrusted network.

export class MeshdbError extends Error {
  constructor(status, message) {
    super(`HTTP ${status}: ${message}`);
    this.name = "MeshdbError";
    this.status = status;
  }
}

export class Client {
  constructor(baseUrl, { user, secret } = {}) {
    this.base = baseUrl.replace(/\/+$/, "");
    this.auth =
      user != null && secret != null
        ? "Bearer " + Buffer.from(`${user}:${secret}`).toString("base64")
        : null;
  }

  async _fetch(method, path, body) {
    const headers = {};
    if (this.auth) headers["Authorization"] = this.auth;
    if (body !== undefined) headers["Content-Type"] = "application/json";
    const res = await fetch(this.base + path, {
      method,
      headers,
      body: body !== undefined ? JSON.stringify(body) : undefined,
    });
    if (!res.ok) {
      let msg = await res.text();
      try {
        msg = JSON.parse(msg).error ?? msg;
      } catch {}
      throw new MeshdbError(res.status, msg);
    }
    return res;
  }

  async _json(method, path, body) {
    return (await this._fetch(method, path, body)).json();
  }

  // -- reads --

  async *query(sql, { shard = 0, params = [], consistency = "linearizable" } = {}) {
    const res = await this._fetch("POST", "/v1/query", { shard, sql, params, consistency });
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buf = "";
    let columns = null;
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });
      let nl;
      while ((nl = buf.indexOf("\n")) >= 0) {
        const line = buf.slice(0, nl).trim();
        buf = buf.slice(nl + 1);
        if (!line) continue;
        const obj = JSON.parse(line);
        if (obj && !Array.isArray(obj) && obj.columns) {
          columns = obj.columns;
          continue;
        }
        if (obj && !Array.isArray(obj) && obj.error) {
          throw new MeshdbError(200, obj.error);
        }
        yield columns ? Object.fromEntries(columns.map((c, i) => [c, obj[i]])) : obj;
      }
    }
  }

  async queryAll(sql) {
    return this._json("POST", "/v1/query_all", { sql });
  }

  async route(key) {
    return (await this._json("POST", "/v1/route", { key })).shard;
  }

  // -- writes --

  async execute(sql, { shard = 0, params = [] } = {}) {
    return this._json("POST", "/v1/execute", { shard, sql, params });
  }

  async tx(statements, { shard = 0 } = {}) {
    const norm = statements.map((s) => (typeof s === "string" ? { sql: s } : s));
    return this._json("POST", "/v1/tx", { shard, statements: norm });
  }

  async executeAll(sql) {
    return this._json("POST", "/v1/execute_all", { sql });
  }

  // -- introspection --

  info() {
    return this._json("GET", "/v1/info");
  }
  cluster() {
    return this._json("GET", "/v1/cluster");
  }
  stats() {
    return this._json("GET", "/v1/stats");
  }
  schema(shard) {
    return this._json("GET", `/v1/schema/${shard}`);
  }
  frames(shard) {
    return this._json("GET", `/v1/frames/${shard}`);
  }

  // -- admin --

  async listUsers() {
    return (await this._json("GET", "/v1/users")).users;
  }
  async createUser(name, secret, role) {
    await this._fetch("POST", "/v1/users", { name, secret, role });
  }
  async dropUser(name) {
    await this._fetch("DELETE", `/v1/users/${name}`);
  }
}

File: drivers/python/meshdb.py
"""meshdb HTTP driver — pure standard library, streaming reads.

A thin client over the meshdb HTTP gateway (`meshdb serve --http ADDR`). No third-party
dependencies: it uses urllib. Queries stream — `query()` is a generator that yields one row
at a time, so a million-row result costs the driver almost nothing, matching the gateway.

    from meshdb import Client
    db = Client("http://localhost:4680", user="app", secret="s3cret")
    for row in db.query("SELECT id, v FROM t WHERE id > ?", params=[5]):
        print(row["id"], row["v"])
    db.execute("INSERT INTO t VALUES (?, ?)", params=[1, "a"])

Auth is sent as `Authorization: Bearer base64(user:secret)` — the programmatic scheme, which
does not trigger a browser login prompt. Over a plaintext gateway the credential is exposed;
run the gateway behind TLS on any untrusted network.
"""

import base64
import json
import urllib.error
import urllib.request


class MeshdbError(Exception):
    def __init__(self, status, message):
        super().__init__(f"HTTP {status}: {message}")
        self.status = status
        self.message = message


class Client:
    def __init__(self, base_url, user=None, secret=None, timeout=30):
        self.base = base_url.rstrip("/")
        self.timeout = timeout
        self._auth = None
        if user is not None and secret is not None:
            token = base64.b64encode(f"{user}:{secret}".encode()).decode()
            self._auth = f"Bearer {token}"

    # -- transport --

    def _open(self, method, path, body=None):
        headers = {}
        if self._auth:
            headers["Authorization"] = self._auth
        data = None
        if body is not None:
            data = json.dumps(body).encode()
            headers["Content-Type"] = "application/json"
        req = urllib.request.Request(self.base + path, data=data, method=method, headers=headers)
        try:
            return urllib.request.urlopen(req, timeout=self.timeout)
        except urllib.error.HTTPError as e:
            raw = e.read().decode(errors="replace")
            msg = raw
            try:
                msg = json.loads(raw).get("error", raw)
            except Exception:
                pass
            raise MeshdbError(e.code, msg) from None

    def _json(self, method, path, body=None):
        with self._open(method, path, body) as resp:
            return json.loads(resp.read())

    # -- reads --

    def query(self, sql, shard=0, params=None, consistency="linearizable"):
        """Stream a read. Yields a dict per row, lazily — the whole result is never held."""
        body = {"shard": shard, "sql": sql, "params": params or [], "consistency": consistency}
        resp = self._open("POST", "/v1/query", body)
        columns = None
        with resp:
            for raw in resp:  # HTTPResponse iterates lines as they arrive
                line = raw.strip()
                if not line:
                    continue
                obj = json.loads(line)
                if isinstance(obj, dict) and "columns" in obj:
                    columns = obj["columns"]
                    continue
                if isinstance(obj, dict) and "error" in obj:
                    # An error after the 200 header, reported as a trailing object.
                    raise MeshdbError(200, obj["error"])
                yield dict(zip(columns, obj)) if columns else obj

    def query_all(self, sql):
        return self._json("POST", "/v1/query_all", {"sql": sql})

    def route(self, key):
        return self._json("POST", "/v1/route", {"key": key})["shard"]

    # -- writes --

    def execute(self, sql, shard=0, params=None):
        return self._json("POST", "/v1/execute", {"shard": shard, "sql": sql, "params": params or []})

    def tx(self, statements, shard=0):
        """Atomic, durable transaction. `statements` is a list of str, (sql, params), or dict."""
        norm = []
        for s in statements:
            if isinstance(s, str):
                norm.append({"sql": s})
            elif isinstance(s, dict):
                norm.append(s)
            else:
                norm.append({"sql": s[0], "params": list(s[1])})
        return self._json("POST", "/v1/tx", {"shard": shard, "statements": norm})

    def execute_all(self, sql):
        return self._json("POST", "/v1/execute_all", {"sql": sql})

    # -- introspection --

    def info(self):
        return self._json("GET", "/v1/info")

    def cluster(self):
        return self._json("GET", "/v1/cluster")

    def stats(self):
        return self._json("GET", "/v1/stats")

    def schema(self, shard):
        return self._json("GET", f"/v1/schema/{shard}")

    def frames(self, shard):
        return self._json("GET", f"/v1/frames/{shard}")

    # -- admin --

    def list_users(self):
        return self._json("GET", "/v1/users")["users"]

    def create_user(self, name, secret, role):
        self._open("POST", "/v1/users", {"name": name, "secret": secret, "role": role}).close()

    def drop_user(self, name):
        self._open("DELETE", f"/v1/users/{name}").close()


# -- Persistent TCP transport (JSON-over-TCP) --

import socket
import struct


class TcpClient:
    """A persistent-connection client over meshdb's JSON-over-TCP protocol.

    Lower per-request overhead than HTTP (no headers, one held socket). One request at a time
    per connection — not safe to share across threads. `query()` streams rows like the HTTP
    client. Auth (name/secret) is sent once at connect; the secret crosses the wire, so run
    over a trusted network or a TLS tunnel.

        db = TcpClient("127.0.0.1", 4620, user="app", secret="s3cret")
        for row in db.query("SELECT id, v FROM t"):
            ...
    """

    def __init__(self, host, port, user=None, secret=None, timeout=30):
        self.sock = socket.create_connection((host, port), timeout=timeout)
        self.sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        self._buf = self.sock.makefile("rb")
        if user is not None and secret is not None:
            r = self._call({"op": "auth", "name": user, "secret": secret})
            if not r.get("ok"):
                raise MeshdbError(401, "authentication failed")

    def close(self):
        try:
            self.sock.close()
        finally:
            self._buf.close()

    def __enter__(self):
        return self

    def __exit__(self, *a):
        self.close()

    def _send(self, frame):
        body = json.dumps(frame).encode()
        self.sock.sendall(struct.pack(">I", len(body)) + body)

    def _recv(self):
        header = self._buf.read(4)
        if len(header) < 4:
            raise MeshdbError(0, "connection closed")
        (n,) = struct.unpack(">I", header)
        return json.loads(self._buf.read(n))

    def _call(self, frame):
        """A bounded op: one request, one result frame."""
        self._send(frame)
        r = self._recv()
        if "error" in r:
            raise MeshdbError(r.get("status", 0), r["error"])
        return r["result"]

    # -- reads --

    def query(self, sql, shard=0, params=None, consistency="linearizable"):
        self._send({"op": "query", "shard": shard, "sql": sql,
                    "params": params or [], "consistency": consistency})
        columns = None
        while True:
            f = self._recv()
            if "columns" in f:
                columns = f["columns"]
            elif "row" in f:
                yield dict(zip(columns, f["row"])) if columns else f["row"]
            elif "end" in f:
                return
            elif "error" in f:
                raise MeshdbError(f.get("status", 200), f["error"])

    def query_all(self, sql):
        return self._call({"op": "query_all", "sql": sql})

    def route(self, key):
        return self._call({"op": "route", "key": key})["shard"]

    # -- writes --

    def execute(self, sql, shard=0, params=None):
        return self._call({"op": "execute", "shard": shard, "sql": sql, "params": params or []})

    def tx(self, statements, shard=0):
        norm = []
        for s in statements:
            if isinstance(s, str):
                norm.append({"sql": s})
            elif isinstance(s, dict):
                norm.append(s)
            else:
                norm.append({"sql": s[0], "params": list(s[1])})
        return self._call({"op": "tx", "shard": shard, "statements": norm})

    def execute_all(self, sql):
        return self._call({"op": "execute_all", "sql": sql})

    # -- introspection & admin --

    def info(self):
        return self._call({"op": "info"})

    def cluster(self):
        return self._call({"op": "cluster"})

    def stats(self):
        return self._call({"op": "stats"})

    def schema(self, shard):
        return self._call({"op": "schema", "shard": shard})

    def frames(self, shard):
        return self._call({"op": "frames", "shard": shard})

    def list_users(self):
        return self._call({"op": "list_users"})["users"]

    def create_user(self, name, secret, role):
        self._call({"op": "create_user", "name": name, "secret": secret, "role": role})

    def drop_user(self, name):
        self._call({"op": "drop_user", "name": name})
